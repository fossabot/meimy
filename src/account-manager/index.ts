export * from './Store';
export * from './Config';
export * from './EimAccount';
export * from './IAccountManagerState';
export * from './IEimAccount';
export * from './RoutePageNames';
