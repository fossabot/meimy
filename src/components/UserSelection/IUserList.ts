
interface IUserList {
    'properties.description'?: string;
    'properties.loginUserName'?: string;
    'properties.displayName'?: string;
    'properties.phoneticName'?: string;
    'properties.lastName'?: string;
    'properties.firstName'?: string;
    'properties.middleName'?: string;
    'properties.birthName'?: string;
    'properties.globalLastName'?: string;
    'properties.globalFirstName'?: string;
    'properties.globalMiddleName'?: string;
    'properties.globalBirthName'?: string;
    'properties.displayGlobalName'?: string;
    'properties.phoneticLastName'?: string;
    'properties.phoneticFirstName'?: string;
    'properties.phoneticMiddleName'?: string;
    'properties.userLocalCompanyName'?: string;
    'properties.userGlobalCompanyName'?: string;
    'properties.faceImage'?: string;
    'properties.officeName'?: string;
    'properties.zip'?: string;
    'properties.location'?: string;
    'properties.country'?: string;
    'properties.region'?: string;
    'properties.state'?: string;
    'properties.city'?: string;
    'properties.street'?: string;
    'properties.building'?: string;
    'properties.mailAddress'?: string;
    'properties.mailAddresses'?: string;
    'properties.phone'?: string;
    'properties.phones'?: string;
    'properties.extension'?: string;
    'properties.extensions'?: string;
    'properties.faxNumber'?: string;
    'properties.language'?: string;
    'properties.timezone'?: string;
    'properties.login'?: string;
    'properties.lockOut'?: string;
    'properties.loginPassword'?: string;
    'properties.passwordExpired'?: string;
    'properties.profile.id'?: string;
    'properties.profile.roleId'?: string;
    'system.documentId'?: string;
    'properties.fullLabel'?: string;
    'properties.label'?: string;
    'properties.departmentIds'?: string;
    'properties.groupIds'?: string;
    'properties.organizationIds'?: string;
    'properties.companyName'?: string;
    'properties.organizationName'?: string;
    'properties.worklocKey'?: string;
    'properties.worklocName'?: string;
    'properties.userType'?: string;
}
