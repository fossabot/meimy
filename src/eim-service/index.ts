export * from './ClientError';
export * from './EIMDocInterface';
export * from './EIMServiceAdapter';
export * from './IDocListForView';
export * from './IDocListSearchOption';
export * from './IResponse';
export * from './IWordResources';
