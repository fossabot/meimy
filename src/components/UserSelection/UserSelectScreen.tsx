import { Body, Button, Form, Icon, Input, Item, List, ListItem, Right, Spinner, Text, Toast, View } from 'native-base';
import React, { Component } from 'react';
import { Modal, Picker, TextStyle, ViewStyle } from 'react-native';

import { getConfig } from '../../account-manager/Config';
import { getEimAccount } from '../../account-manager/EimAccount';
import { getDocListValue } from '../../eim-service/EIMServiceAdapter';
import { IDocListRowForView } from '../../eim-service/IDocListForView';
import { IDocListSearchOption, SearchCondition } from '../../eim-service/IDocListSearchOption';
import { DirectoryTypeKey } from '../UserBadge';
import { IGroupList, IOrganizationList, IUserList } from './IUserList';

const config = getConfig();

type DocListName = 'userdoclist' | 'groupdoclist' | 'organizationdoclist';

type directoryTypeName = 'ユーザー' | '組織' | 'グループ';

const directoryTypeKeys: DirectoryTypeKey[] = [
    'user',
    'group',
    'organization',
];

const directoryType: { [key in DirectoryTypeKey]: directoryTypeName } = {
    user: 'ユーザー',
    group: 'グループ',
    organization: '組織',
};

export interface IFilter {
    users: boolean;
    groups: boolean;
    organizations: boolean;
}

interface IOptionalProps {
    filter: IFilter;
}

interface IProps extends Partial<IOptionalProps> {
    onSelect: (docId: string, type: DirectoryTypeKey) => void;
}


interface ISearchedListItem {
    docId: string;
    displayName: string;
    corpName: string;
    orgName: string;
    faceImageId: string;
}

interface IState {
    selectedDirectoryType: DirectoryTypeKey;
    searchResult: ISearchedListItem[];
    searchWords: string;
    searchCondition: {
        offset: number;
        limit: number;
    };
    shown: boolean;
    processing: boolean;
}

const containerStyle: ViewStyle = {
    backgroundColor: '#fff',
    width: '95%',
    height: '95%',
};

const searchBox: ViewStyle = {
    flexDirection: 'row',
};

const listOrgStyle: TextStyle = {
    fontSize: 12,
    color: config.colorPalets.$frontDisabledColor,
};

export class UserSelectScreen extends Component<IProps, IState> {
    public static defaultProps: IOptionalProps = {
        filter: {
            users: true,
            groups: true,
            organizations: true,
        }
    }
    private searchedDirType: DirectoryTypeKey = 'user';
    private searchedWord: string = '';
    private getInitState = (): IState => ({
        searchResult: [],
        selectedDirectoryType: 'user',
        searchWords: '',
        searchCondition: {
            limit: 30,
            offset: 0,
        },
        shown: false,
        processing: false,
    });
    public constructor(props: IProps) {
        super(props);
        this.state = this.getInitState();
    }
    public render() {
        return (
            <Modal transparent animated
                animationType="fade" visible={this.state.shown}>
                <View style={{
                    flex: 1,
                    backgroundColor: 'rgba(0,0,0,0.5)',
                    alignItems: 'center',
                    justifyContent: 'center'
                }}>
                    <View style={containerStyle}>
                        <View style={{ flexDirection: 'row-reverse' }}>
                            <Button key="close-button" icon transparent
                                onPress={this.closeButtonPress}>
                                <Icon name="close" />
                            </Button>
                        </View>
                        <View>
                            <Form style={searchBox}>
                                <Item fixedLabel style={{ flexGrow: 1 }}>
                                    <Input key="search-word-input" value={this.state.searchWords}
                                        placeholder="検索"
                                        onChangeText={this.changeSearchWords}
                                        returnKeyType="search" />
                                </Item>
                                <Button key="search-button"
                                    icon transparent style={{ flexGrow: 0 }}
                                    onPress={this.pressSearchButton}>
                                    <Icon name="search" />
                                </Button>
                                <View style={{ flexGrow: 1 }}>
                                    {this.createDirectoryTypePicker(this.state.selectedDirectoryType)}
                                </View>
                            </Form>
                            <List key="result-list">
                                {this.createSearchedUserList()}
                            </List>
                            {
                                // 更に表示 ボタン
                                (this.state.processing) ?
                                    <Spinner /> :
                                    (0 < this.state.searchResult.length) ?
                                        <Button key="more-search-button" full
                                            onPress={this.pressMoreSearch}>
                                            <Text>さらに表示</Text>
                                        </Button> :
                                        null
                            }
                        </View>
                    </View>
                </View>
            </Modal>
        )
    }
    public show = () => {
        const s = this.getInitState();
        s.shown = true;
        this.setState(s);
    }
    private closeButtonPress = () => {
        this.setState({
            shown: false,
        });
    }
    private changeSearchWords = (value: string) => {
        this.setState({
            searchWords: value,
        });
    }
    private createSearchedUserList = () => {
        return this.state.searchResult.map(item => (
            <ListItem key={item.docId}>
                <Body>
                    <Text>{item.displayName}</Text>
                    <Text style={listOrgStyle}>{item.orgName}</Text>
                </Body>
                <Right>
                    <Button icon transparent success
                        onPress={this.pressResultRow.bind(this, item.docId)}>
                        <Icon name="md-add-circle" />
                    </Button>
                </Right>
            </ListItem>
        ));
    }
    private pressResultRow = (docId: string) => {
        this.props.onSelect(docId, this.searchedDirType);
        this.closeButtonPress();
    }
    private createDirectoryTypePicker = (selectedDirectoryType: DirectoryTypeKey) => {
        const items: JSX.Element[] =
            directoryTypeKeys.map(key => (
                <Picker.Item key={key} label={directoryType[key]} value={key} />
            ));
        return (
            <Picker
                key="dir-picker"
                mode="dropdown"
                selectedValue={selectedDirectoryType}
                onValueChange={this.changeDirectoryType.bind(this)}>
                {items}
            </Picker>);
    }

    private changeDirectoryType = (key: DirectoryTypeKey) => {
        this.setState({
            selectedDirectoryType: key,
        });
    }
    private pressMoreSearch = () => {
        this.setState({
            processing: true,
        });
        this.startSearch[this.state.selectedDirectoryType]();
    }
    private pressSearchButton = () => {
        this.searchedWord = this.state.searchWords;
        this.setState({
            searchResult: [],
            searchCondition: {
                limit: this.state.searchCondition.limit,
                offset: 0,
            },
            processing: true,
        });
        this.startSearch[this.state.selectedDirectoryType]();
    }
    private startSearch: { [key in DirectoryTypeKey]: () => Promise<void> } = {
        group: () => {
            const search = this.createSearchCondition<IGroupList>([{
                ignoreCaseSense: false,
                operator: 'equal',
                propertyName: 'properties.groupType',
                type: 'string',
                value: 'group',
            }]);
            const condition: IDocListSearchOption<IGroupList> = {
                ...this.state.searchCondition,
                search,
            };
            return this.commonSearch('groupdoclist',
                this.createGroupRowData, condition);
        },
        organization: () => {
            const search = this.createSearchCondition<IGroupList>();
            const condition: IDocListSearchOption<IGroupList> = {
                ...this.state.searchCondition,
                search,
            };
            return this.commonSearch('organizationdoclist',
                this.createOrganizationRowData, condition);
        },
        user: () => {
            const search = this.createSearchCondition<IUserList>([
                '(',
                {
                    propertyName: 'properties.userType',
                    type: 'string',
                    operator: 'equal',
                    value: 'user',
                    ignoreCaseSense: false
                },
                'or',
                {
                    propertyName: 'properties.userType',
                    type: 'string',
                    operator: 'equal',
                    value: null,
                    ignoreCaseSense: false
                },
                ')',]);
            const condition: IDocListSearchOption<IUserList> = {
                ...this.state.searchCondition,
                search,
            };
            return this.commonSearch('userdoclist',
                this.createUserRowData,
                condition);
        }
    }
    private createSearchCondition = <T extends {}>(searches?: SearchCondition<T>[]): SearchCondition<T>[] => {
        const argConditons: SearchCondition<T>[] = !searches ? [] : ['(', ...searches, ')', 'and'];
        const fulltext: SearchCondition<T> = {
            ignoreCaseSense: true,
            propertyName: 'fulltextsearch_list',
            type: 'string',
            operator: 'fulltextsearch_sentence',
            value: this.searchedWord,
        };
        return [...argConditons, fulltext];
    }
    private commonSearch = async <T extends {}>(listName: DocListName,
        createRow: (item: IDocListRowForView<T>) => ISearchedListItem,
        condition: IDocListSearchOption<T>) => {
        this.searchedDirType = this.state.selectedDirectoryType;
        const account = getEimAccount();
        const sa = account.getServiceAdapter();
        try {
            const result = await sa.getDocListForView<T>(
                account.eimTokens,
                'addressbook',
                listName,
                {
                    ...condition,
                });
            const addList = result.docList.map(createRow);
            const newList = this.state.searchResult.concat(addList);
            condition.offset += result.docList.length;
            this.setState({
                searchResult: newList,
                searchCondition: condition,
                processing: false,
            });
        } catch {
            Toast.show({
                text: 'ネットワーク エラー',
                type: 'danger',
            });
        }
    }
    private createUserRowData = (item: IDocListRowForView<IUserList>): ISearchedListItem => {
        const { columnValues } = item;
        return {
            corpName: (getDocListValue<IUserList>(columnValues, 'properties.companyName') as string) || '',
            displayName: (getDocListValue<IUserList>(columnValues, 'properties.displayName') as string) || '',
            faceImageId: '',
            docId: item.documentId,
            orgName: (getDocListValue<IUserList>(columnValues, 'properties.organizationName') as string) || '',
        };
    }
    private createGroupRowData = (item: IDocListRowForView<IGroupList>): ISearchedListItem => {
        const { columnValues } = item;
        return {
            corpName: '',
            displayName: (getDocListValue<IGroupList>(columnValues, 'properties.fullLabel') as string) || '',
            faceImageId: '',
            docId: item.documentId,
            orgName: '',
        };
    }
    private createOrganizationRowData = (item: IDocListRowForView<IOrganizationList>): ISearchedListItem => {
        const { columnValues } = item;
        return {
            corpName: '',
            displayName: (getDocListValue<IOrganizationList>(columnValues, 'properties.fullLabel') as string) || '',
            faceImageId: '',
            docId: item.documentId,
            orgName: '',
        };
    }
}

export default UserSelectScreen
