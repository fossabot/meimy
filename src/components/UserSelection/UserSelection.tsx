import { Button, Icon, View } from 'native-base';
import React, { Component } from 'react';
import { ViewStyle } from 'react-native';

import { DirectoryTypeKey, IUserBadgeOptionalProps, IUserBadgeProps, UserBadge } from '../UserBadge';
import UserSelectScreen, { IFilter } from './UserSelectScreen';

interface IOptionalProps {
    addable: boolean;
    filter?: IFilter;
    multiple: boolean;
    style?: ViewStyle;
    userBadgeProp: { [key in keyof IUserBadgeOptionalProps]?: IUserBadgeOptionalProps[key] };
}
export interface IUserListItem {
    type: DirectoryTypeKey;
    userDocId: string;
}
export interface IUserSelectionProps extends Partial<IOptionalProps> {
    onChange: (userIds: IUserListItem[]) => void;
    selectedUsers: IUserListItem[];
    showAddButton: boolean;
}

const containerState: ViewStyle = {
    flexDirection: 'row',
    flexWrap: 'wrap',
};

export class UserSelection extends Component<IUserSelectionProps> {
    public static defaultProps: IOptionalProps = {
        addable: false,
        multiple: false,
        userBadgeProp: {},
    }
    private selectionModal: UserSelectScreen | null = null;
    public constructor(props: IUserSelectionProps) {
        super(props);
    }
    public render() {
        const { props } = this;
        const style = Object.assign({}, containerState, props.style);
        return (
            <View style={style}>
                {this.createUserList()}
                {
                    this.props.showAddButton ?
                        <Button transparent icon success
                            key="add-button"
                            onPress={this.addButtonPress}>
                            <Icon name="md-add-circle" />
                        </Button> : null
                }
                <UserSelectScreen
                    key="user-select-screen"
                    ref={(me) => this.selectionModal = me}
                    filter={props.filter}
                    onSelect={this.addList} />
            </View>
        );
    };
    private addList = (userDocId: string, type: DirectoryTypeKey) => {
        const existUser = this.props.selectedUsers.find(u => u.userDocId === userDocId);
        if (existUser && this.props.multiple) {
            return;
        }
        const newItem = {
            type,
            userDocId,
        };
        const userList: IUserListItem[] =
            (this.props.multiple) ?
                [...this.props.selectedUsers, newItem] :
                [newItem];
        // すでに有れば無視
        this.onChange(userList);
    }
    private onChange = (userList: IUserListItem[]) => {
        this.props.onChange(userList);
    }
    private addButtonPress = () => {
        if (!!this.selectionModal) { this.selectionModal.show(); };
    }
    private createUserList = () => {
        const { props } = this;
        if (props.multiple) {
            return this.createUserListForArray(props.selectedUsers);
        } else {
            return this.createUserListForArray(props.selectedUsers.slice(0, 1));
        }
    }
    private createUserListForArray = (users: IUserListItem[]) => {
        return users.map(user => {
            const defaultStyle: ViewStyle = {
                margin: 4,
            };
            const userProps = (this.props as IOptionalProps).userBadgeProp;
            if (!!userProps.style) {
                userProps.style = Object.assign({}, defaultStyle, userProps.style);
            } else {
                userProps.style = defaultStyle;
            }
            const props: IUserBadgeProps = {
                ...userProps,
                userId: user.userDocId,
                onLongPress: this.onDelete,
            };
            return <UserBadge key={props.userId} {...props} />
        });

    }
    private onDelete = (userDocId: string) => {
        const { selectedUsers } = this.props;
        const index = selectedUsers.findIndex(a => a.userDocId === userDocId);
        if (index === -1) { return; }
        selectedUsers.splice(index, 1);
        this.onChange([...selectedUsers]);
    }
}

export default UserSelection
