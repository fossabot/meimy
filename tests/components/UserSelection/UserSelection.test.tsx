import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';

import { IUserBadgeProps } from '../../../src';
import UserSelection from '../../../src/components/UserSelection/UserSelection';

Enzyme.configure({
    adapter: new EnzymeAdapter(),
});

const getUsers = (): IUserBadgeProps[] => {
    return [
        {
            badgeColor: '#333',
            textColor: '#ccc',
            type: 'user',
            userId: 'a0001',
            style: {
                borderColor: '#030',
            }
        },
        {
            type: 'group',
            userId: 'a0002',
        },
    ];
};

describe('init', () => {
    test('', () => {
        const users = getUsers();
        const target = Enzyme.shallow(
            <UserSelection selectedUsers={users}
                badgeColor='#900' textColor="#099"
                style={{ backgroundColor: '#100' }}
                userBadgeStyle={{ borderColor: '#111', borderWidth: 1 }}
            />);
        expect(toJson(target)).toMatchSnapshot();
    });
});
