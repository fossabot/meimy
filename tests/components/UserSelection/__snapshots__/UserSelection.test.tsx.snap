// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init  1`] = `
<Styled(ViewNB)
  style={
    Object {
      "backgroundColor": "#100",
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <UserBadge
    badgeColor="#333"
    color="#fff"
    key="a0001"
    style={
      Object {
        "borderColor": "#030",
        "borderWidth": 1,
        "margin": 4,
      }
    }
    textColor="#ccc"
    type="user"
    userId="a0001"
  />
  <UserBadge
    badgeColor="#900"
    color="#fff"
    key="a0002"
    style={
      Object {
        "borderColor": "#111",
        "borderWidth": 1,
        "margin": 4,
      }
    }
    textColor="#099"
    type="group"
    userId="a0002"
  />
  <Styled(Button)
    icon={true}
    success={true}
    transparent={true}
  >
    <Styled(Icon)
      name="md-add-circle"
    />
  </Styled(Button)>
</Styled(ViewNB)>
`;
